testing-pre-releases

fabriziopandini opened the pull request:
This PR documents how to test pre-release versions of Kubernetes with kubeadm; which version numbers I should take care of, where to retrieve pre-compiled artefacts, how to build locally.

/CC @kubernetes/sig-cluster-lifecycle-pr-reviews 

This PR is part of #849 checklist

timothysc commented:
/assign @liztio

neolit123 commented on CONTRIBUTING.md:
```diff
@@ -37,6 +37,10 @@ However, you do not frequently have to build all of kubernetes to work on kubead
 
+### Testing pre-release versions of Kubernetes with kubeadm
+
+See [testing-pre-releases.md](testing-pre-releases.md) for information about how to get pre-release versions of kubernetes or kubeadm and how to test it.
```
`s/test it/test them/`

liztio replied:
There's plenty of information on how to get these prerelease versions, but not much on testing. What would a deployment look like? How can you test say, a pre-release kubelet binary? What verification steps should be run?

neolit123 commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+
+As alternative, you can browse <https://console.cloud.google.com/gcr/images/{gcr-repository-name}/GLOBAL/{image-name}>
+
+To retrieve a pre-compiled version of control plane images `docker pull {iamge tag}`
```
`s/iamge/image`

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+  it is convenient to include also kubeadm package in this group to get the kubelet drop-in file deployed automatically.
+- The version of the kubeadm binary to be used, that can eventually overwrite the kubeadm binary included in
+  the .deb or .rpm packages. e.g. Use v1.10.3 GA packages for kubelet, kubectl, but test a local build of kubernetes.
+- The `--kubernetes-version` flag of kubeadm or the `kubernetesVersion` field of Kubeadm API,
```
maybe specify this as the config file, and link to the config file docs

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+- The `--kubernetes-version` flag of kubeadm or the `kubernetesVersion` field of Kubeadm API,
+  defining the version number of the control plane components installed by kubeadm.
+
+For each of the above set version number, it is also necessary to understand the supported version skew policy,
```
What exactly does this look like? It's hard to suss out the actions you should be taking based on this.

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+The [Kubeadm version skew policy](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/#version-skew-policy)
+can briefly be summarized in:
+
+- The kubeadm CLI tool of version vX.Y may deploy clusters with a control plane of version vX.Y or vX.(Y-1).
```
This isn't true anymore is it? I thought vX.Y can only deploy vX.Y

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+
+- The kubeadm CLI tool of version vX.Y may deploy clusters with a control plane of version vX.Y or vX.(Y-1).
+- The kubeadm CLI vX.Y can upgrade an existing kubeadm-created cluster of version vX.(Y-1).
+- One minor version skew between the kubelet and the control plane is supported, but the kubelet version may never
```
might want to spell out major.minor.patch, I always get the latter two confused.

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+
+|                         | .deb or .rpm                                                 | kubeadm binary                                               | control plane                                                |
+| ----------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| **GA release**          | from .deb or .rpm repository                                 | from [gs://kubernetes-release/release/](gs://kubernetes-release/release/) | from `gcr.io/google-containers`                              |
```
might want to document the `gs` protocol / how we can get access to it

liztio commented on testing-pre-releases.md:
````diff
@@ -0,0 +1,264 @@
+To explore versions available in Google Storage buckets use:
+
+```bash
+gsutil ls {bucket-name}{filter}
````
could you link to where you get the `gsutil` binary?

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+gsutil ls {bucket-name}{filter}
+
+# e.g. search all GA and alpha/beta v1.10 releases
+gsutil ls -g gs://kubernetes-release/release/v1.10*
```
```
$ gsutil ls -g 'gs://kubernetes-release/release/v1.10*'
CommandException: Incorrect option(s) specified. Usage:

  gsutil ls [-a] [-b] [-d] [-l] [-L] [-r] [-p proj_id] url...

For additional help run:
  gsutil help ls
zsh: exit 1     gsutil ls -g 'gs://kubernetes-release/release/v1.10*'
```

liztio replied:
looks like `-d` might be what you want?

liztio commented on testing-pre-releases.md:
````diff
@@ -0,0 +1,264 @@
+To retrieve a pre-compiled version of kubeadm binary use:
+
+```bash
+gsutil cp {bucket-name}/{release}/bin/linux/arm64/kubeadm .
````
you need the `gs://` for this to work properly

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+- prefix `ci/` followed by Kubernetes release labels for CI/CD versions like e.g. `ci/latest`, `ci/latest-1`, `latest-1.10`.
+  See <https://console.cloud.google.com/storage/browser/kubernetes-release-dev/ci-cross/> for the full list of labels.
+
+Instead, if you want to retrieve manually pre-compiled/pre-built GA, alpha/beta versions of control plane images, such
```
maybe nix this "instead"

liztio commented on testing-pre-releases.md:
````diff
@@ -0,0 +1,264 @@
+To explore versions available in Google Container Registry use
+
+```bash
+gcloud container images list-tags gcr.io/{gcr-repository-name}/{image-name} --sort-by=~tags --filter=tags:v1.10.2 --limit=50
````
it's kind of weird that some of the fields are placeholders (iimage name) but tags is already filled in. Maybe have one example with all fields empty and another with all fields full?

liztio commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,264 @@
+
+build output will be stored in `bazel-bin/build/debs`.
+
+> cross build not supported yet; the unofficial Planter tool can be used to overcome the problem
```
Link to Planter?

fabriziopandini commented:
@stealthybox 👍 

@liztio all comment are address
- for the version skew policy, this is what it is on kubernetes.io now and I don't recall of recent changes 
- I added a "Creating the Kubernetes cluster with kubeadm" paragraph that highlights changes to the standard procedure when using pre-releases
- I added a Testing the Kubernetes cluster with both a "Smoke test" and conformance tests with Heptio Sonobuoy

detiber commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,530 @@
+
+- The kubeadm CLI tool of version MAJOR.MINOR may deploy clusters with a control plane of version MAJOR.MINOR or MAJOR.(MINOR-1).
+- The kubeadm CLI MAJOR.MINOR can upgrade an existing kubeadm-created cluster of version MAJOR.(MINOR-1).
+- One minor version skew between the kubelet and the control plane is supported, but the kubelet version may never
```
s/may never/must not/

detiber commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,530 @@
+The [Kubeadm version skew policy](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/#version-skew-policy)
+can briefly be summarized in:
+
+- The kubeadm CLI tool of version MAJOR.MINOR may deploy clusters with a control plane of version MAJOR.MINOR or MAJOR.(MINOR-1).
```
Would it be worth mentioning the PATCH version as part of this?

fabriziopandini replied:
@liztio @detiber I'm going to remove the details of the version skew here and leave only the links to "official" version skew policy in Kubernestes.io.
If something should be improved or fixed, let's do this on the master copy

detiber commented on testing-pre-releases.md:
```diff
@@ -0,0 +1,530 @@
+|                         | .deb or .rpm                                                 | kubeadm binary                                               | control plane                                                |
+| ----------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| **GA release**          | from .deb or .rpm repository                                 | from `gs://kubernetes-release/release/` GCS bucket | from `gcr.io/google-containers` container registry           |
+| **alpha/beta release*** | not available. use CI/CD version "near" tag                  | from `gs://kubernetes-release/release/` GCS bucket | from `gcr.io/google-containers` container registry           |
```
Instead of mentioning the gs bucket for GA/alpha/beta releases, should we just mention the github release instead?

fabriziopandini commented:
@detiber Added "retrieve from github release page" option (nb. afaik this option doesn't cover all the possible combinations, so also other options are still there)
@liztio @detiber version skew now links to kubernetes.io only

timothysc commented:
/lgtm
